From ddd143b4c60569064e8633618189776b928ea267 Mon Sep 17 00:00:00 2001
Message-Id: <ddd143b4c60569064e8633618189776b928ea267.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Gaurav K Singh <gaurav.k.singh@intel.com>
Date: Thu, 25 Dec 2014 01:47:56 +0530
Subject: [PATCH 21/95] MUST_REBASE [VPG]: video/adf/intel: GPIO changes
 required for CHV

CHV GPIO configuration is different from VLV. There are a total of
284 pins and the pin no varies from 0 to 283. In order to genealize the
GPIO programming, all these pins have been categorised into four communities.
Depending on the pin no, corresponding community is selected and
GPIO is programmed accordingly.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

Issue: GMINL-5545
Change-Id: If39981aab112fb3e5758b5dcc9e92458c0d9058c
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
---
 .../core/common/dsi/panels/intel_dsi_panel_vbt.c   |   66 +++++++++++++++++++-
 .../video/adf/intel/include/core/vlv/vlv_dc_regs.h |   19 ++++++
 2 files changed, 82 insertions(+), 3 deletions(-)

diff --git a/drivers/video/adf/intel/core/common/dsi/panels/intel_dsi_panel_vbt.c b/drivers/video/adf/intel/core/common/dsi/panels/intel_dsi_panel_vbt.c
index c18c9c2..f43aefe 100644
--- a/drivers/video/adf/intel/core/common/dsi/panels/intel_dsi_panel_vbt.c
+++ b/drivers/video/adf/intel/core/common/dsi/panels/intel_dsi_panel_vbt.c
@@ -303,11 +303,71 @@ static u8 *mipi_exec_delay(struct dsi_pipe *dsi_pipe, u8 *data)
 	return data;
 }
 
+static u8 *chv_mipi_exec_gpio(struct dsi_pipe *dsi_pipe, u8 *data)
+{
+	u8 gpio, action;
+	u32 function, pad;
+	u32 port;
+	u16 familyno, padno;
+
+	/*
+	 * Skipping the first byte as it is of no
+	 * interest for android in new version
+	 */
+	if (dsi_pipe->config.dsi->seq_version >= 3)
+		data++;
+
+	gpio = *data++;
+
+	/* pull up/down */
+	action = *data++;
+
+	if (gpio > (TOTAL_NO_OF_GPIOS - 1)) {
+		pr_err("%s: Unsupported gpio index for CHV\n", __func__);
+		return NULL;
+	}
+	if (gpio >= 0 && gpio <= (NO_OF_NORTH_CORE_GPIOS - 1))
+		port = IOSF_PORT_GPIO_NC;
+	else if (gpio >= NO_OF_NORTH_CORE_GPIOS && gpio <=
+		((NO_OF_NORTH_CORE_GPIOS + NO_OF_EAST_CORE_GPIOS) - 1)) {
+		port = IOSF_PORT_GPIO_EC;
+		gpio = gpio - NO_OF_NORTH_CORE_GPIOS;
+	} else if (gpio >= (NO_OF_NORTH_CORE_GPIOS + NO_OF_EAST_CORE_GPIOS) &&
+		gpio <= ((NO_OF_NORTH_CORE_GPIOS + NO_OF_EAST_CORE_GPIOS +
+		NO_OF_SOUTH_WEST_CORE_GPIOS) - 1)) {
+		port = IOSF_PORT_GPIO_SWC;
+		gpio = gpio - (NO_OF_NORTH_CORE_GPIOS + NO_OF_EAST_CORE_GPIOS);
+	} else if (gpio >= (NO_OF_NORTH_CORE_GPIOS + NO_OF_EAST_CORE_GPIOS +
+		NO_OF_SOUTH_WEST_CORE_GPIOS &&
+		gpio <= (TOTAL_NO_OF_GPIOS - 1))) {
+		port = IOSF_PORT_GPIO_SEC;
+		gpio = gpio - (NO_OF_NORTH_CORE_GPIOS + NO_OF_EAST_CORE_GPIOS +
+		NO_OF_SOUTH_WEST_CORE_GPIOS);
+	}
+
+	familyno = gpio / MAX_FAMILY_PAD_GPIO_NO;
+	padno = gpio % MAX_FAMILY_PAD_GPIO_NO;
+	pad = FAMILY0_PAD_REGS_BASE + (FAMILY_PAD_REGS_SIZE * familyno) +
+			(GPIO_REGS_SIZE * padno) + PAD_CONF0_OFFSET;
+	function = pad + PAD_CONF1_OFFSET;
+	vlv_gpio_write(port, function, 0x00000000);
+	vlv_gpio_write(port, pad, 0x00008100 |
+				(action << 1));
+
+	return data;
+}
+
 static u8 *mipi_exec_gpio(struct dsi_pipe *dsi_pipe, u8 *data)
 {
 	u8 gpio, action;
-	u16 function, pad;
-	u32 val, port;
+	u32 function, pad;
+	u32 port;
+	u32 val;
+
+	if (IS_CHERRYVIEW()) {
+		data = chv_mipi_exec_gpio(dsi_pipe, data);
+		goto out;
+	}
 
 	/*
 	 * Skipping the first byte as it is of no
@@ -350,7 +410,7 @@ static u8 *mipi_exec_gpio(struct dsi_pipe *dsi_pipe, u8 *data)
 
 	/* pull up/down */
 	vlv_gpio_write(port, pad, val);
-
+out:
 	return data;
 }
 
diff --git a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
index aa0658b..350ffc4 100644
--- a/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
+++ b/drivers/video/adf/intel/include/core/vlv/vlv_dc_regs.h
@@ -396,6 +396,25 @@ extern int chv_cursor_offsets[];
 #define VLV_IOSF_DATA				(VLV_DISPLAY_BASE + 0x2104)
 #define VLV_IOSF_ADDR				(VLV_DISPLAY_BASE + 0x2108)
 
+
+#define TOTAL_NO_OF_GPIOS			284
+#define NO_OF_NORTH_CORE_GPIOS			73
+#define NO_OF_EAST_CORE_GPIOS			27
+#define NO_OF_SOUTH_WEST_CORE_GPIOS		98
+#define NO_OF_SOUTH_EAST_CORE_GPIOS		86
+
+#define GPIO_REGS_SIZE				8
+#define MAX_FAMILY_PAD_GPIO_NO			15
+#define FAMILY0_PAD_REGS_BASE			0x4400
+#define FAMILY_PAD_REGS_SIZE			0x400
+
+#define PAD_CONF0_OFFSET			0
+#define PAD_CONF1_OFFSET			4
+
+#define IOSF_PORT_GPIO_EC			0xA8
+#define IOSF_PORT_GPIO_SEC			0x48
+#define IOSF_PORT_GPIO_SWC			0xB2
+
 /* See configdb bunit SB addr map */
 #define BUNIT_REG_BISOC				0x11
 
-- 
1.7.9.5

